packages/server-wallet/src/wallet/store.ts

```diff
@@ -504,13 +500,23 @@ export class Store {
       .returning('*')
       .first();
 
-    return ObjectiveModel.insert(objectiveToBeStored, tx) as Promise<DBOpenChannelObjective>;
+    return this.saveObjective(objectiveToBeStored, tx);
+  }
+
+  private async saveObjective(
+    objective: ObjectiveToStore,
+    tx: Transaction
+  ): Promise<EnsureObjectiveResult> {
+    const objectiveId = calculateObjectiveId(objective);
+    const isNew = !(await ObjectiveModel.doesExist(objectiveId, tx));
+    const storedObjective = await ObjectiveModel.insert(objective, tx);
```

Comment:
I don't really think this is the best pattern.
1. It's two queries, which is inefficient.
2. It's not obviously correct.

though I think (2) might be wrong, it might obviously be correct?


Did you consider turning `insert` into an `upsert`? It can return a value that indicates that the row was already present, something like
```
INSERT INTO blah
VALUES (foo)
ON CONFLICT (...) DO NOTHING
RETURNING *, trueWhenRowDidExist
```

Reply:
There is a relatively succinct hack [here](https://stackoverflow.com/questions/34708509/how-to-use-returning-with-on-conflict-in-postgresql) whose tradeoff might not apply:
```
INSERT INTO blah
VALUES (foo)
ON CONFLICT (...) DO UPDATE SET id = excluded.id
RETURNING *, trueWhenRowDidExist
```

I don't know how to compute `trueWhenRowDidExist` yet. You could compare `existing.created_at` with `now()`, which might result in a rare false positive.

Reply:
> I don't really think this is the best pattern.
> 
> 1. It's two queries, which is inefficient.
> 2. It's not obviously correct.
> 
> though I think (2) might be wrong: it might obviously be correct within a transaction, but it might throw a surprising error (which this PR does not seem to account for)

I think it's good enough to be correct in the transaction? Shouldn't any error thrown in a transaction cause us to rollback? Is there a particular error case you're thinking about?

> Did you consider turning `insert` into an `upsert`? It can return a value that indicates that the row was already present, something like
> 
> ```
> INSERT INTO blah
> VALUES (foo)
> ON CONFLICT (...) DO NOTHING
> RETURNING *, trueWhenRowDidExist
> ```

I went with a separate query as it seemd more straightforward than using an `upsert`. (I would argue all the caveats mentioned [here](https://stackoverflow.com/questions/34708509/how-to-use-returning-with-on-conflict-in-postgresql) are a point in that favor)

That said we might want to go with `upsert` for performance reasons.